azimuth_schedule_operator/operator.py

```diff
@@ -120,3 +189,339 @@ async def schedule_check(body, namespace, **_):
+        lease.status.set_phase(lease_crd.LeasePhase.PENDING)
+        await save_instance_status(lease)
+
+    # Create a cloud instance from the referenced credential secret
+    secrets = await K8S_CLIENT.api("v1").resource("secrets")
+    cloud_creds = await secrets.fetch(
+        lease.spec.cloud_credentials_secret_name, namespace=lease.metadata.namespace
+    )
+    async with openstack.from_secret_data(cloud_creds.data) as cloud:
+        # If the lease has no end date, we don't attempt to use Blazar
+        if not lease.spec.ends_at:
+            logger.info("lease has no end date")
+            await update_lease_status_no_blazar(cloud, lease)
+            await save_instance_status(lease)
+            return
+
+        # If the lease has an end date, we might need to do some Blazar stuff
+        if blazar_enabled(cloud):
+            blazar_client = cloud.api_client("reservation", timeout=30)
+            logger.info("checking if blazar lease exists")
+            blazar_lease_name = f"az-{lease.metadata.name}"
+            blazar_lease = await find_blazar_lease(blazar_client, blazar_lease_name)
+            if not blazar_lease:
+                # NOTE(mkjpryor)
+                #
+                # We only create the Blazar lease if we are in the PENDING phase
+                #
+                # If we are in any other phase and the lease does not exist, then we
+                # leave the status as-is but log it. This can happen in one of three
+                # ways:
+                #
+                #  1. The lease was not created due to an unrecoverable error
+                #  2. The lease we created was deleted by someone else
+                #  3. Blazar has been enabled on a cloud after a lease has already been
+                #     processed using the non-Blazar code path
+                if lease.status.phase == lease_crd.LeasePhase.PENDING:
+                    logger.info("creating blazar lease")
+                    try:
+                        blazar_lease = await create_blazar_lease(
+                            blazar_client, blazar_lease_name, lease
+                        )
+                    except BlazarLeaseCreateError as exc:
+                        logger.error(str(exc))
+                        lease.status.set_phase(lease_crd.LeasePhase.ERROR, str(exc))
+                        await save_instance_status(lease)
+                        return
+                else:
+                    phase = lease.status.phase.name
+                    logger.warn(f"phase is {phase} but blazar lease does not exist")
+            # Set the status from the created lease
+            if blazar_lease:
+                blazar_lease_status = blazar_lease["status"]
+                logger.info(f"blazar lease has status '{blazar_lease_status}'")
+                lease.status.set_phase(lease_crd.LeasePhase[blazar_lease_status])
+                if lease.status.phase == lease_crd.LeasePhase.ACTIVE:
```

Comment:
I would be tempted not to do that here, and wait for the timer to deal with it for all code paths.